jupyterhub/tests/test_api.py

```diff
@@ -13,6 +13,7 @@
 import requests
 
 from tornado import gen
+from tornado.concurrent import Future as tFuture
```

Comment:
Why is the tornado `Future` different from the `concurrent.futures` one? If I use `tFuture` in the mock below it works, but with `Future` I get an error that `object of type Future can't be await'ed`.

(this needs to be made nicer before merging)

Reply:
There are three kinds of Future:

- asyncio.Future (The New Way)
- tornado.Future (with tornado 5, these are really asyncio.Future)
- concurrent.futures.Future (The Slightly Less New Way) \*

concurrent Futures are not awaitable, so must be wrapped in `asyncio.wrap_future`. Why? I don't know, but that's how it is. Our `utils.maybe_future` wrapper turns all of these things (and more!) into awaitable Futures.

\* - *multiprocessing.AsyncResult (The Old Way)*

Reply:
Ah ok. Will switch the mock to returning a `maybe_future`. Thanks for the history lesson.